pkg/storage/couchdb/couchdbstore.go

```diff
@@ -0,0 +1,130 @@
+/*
+Copyright SecureKey Technologies Inc. All Rights Reserved.
+SPDX-License-Identifier: Apache-2.0
+*/
+
+package couchdbstore
+
+import (
+	"context"
+	"encoding/json"
+	"errors"
+
+	_ "github.com/go-kivik/couchdb" // The CouchDB driver
+	"github.com/go-kivik/kivik"
+
+	"github.com/trustbloc/edge-core/pkg/storage"
+)
+
+// Provider represents an CouchDB implementation of the storage.Provider interface
+type Provider struct {
+	hostURL       string
+	couchDBClient *kivik.Client
+}
+
+const blankHostErrMsg = "hostURL for new CouchDB provider can't be blank"
+
+// NewProvider instantiates Provider
+func NewProvider(hostURL string) (*Provider, error) {
+	if hostURL == "" {
+		return nil, errors.New(blankHostErrMsg)
+	}
+
+	client, err := kivik.New("couch", hostURL)
+	if err != nil {
+		return nil, err
+	}
+
+	return &Provider{hostURL: hostURL, couchDBClient: client}, nil
+}
+
+// OpenStore opens and returns a store for the given name.
+func (p *Provider) OpenStore(name string) (storage.Store, error) {
+	exists, err := p.couchDBClient.DBExists(context.Background(), name)
+	if err != nil {
+		return nil, err
+	}
+
+	if exists {
+		db := p.couchDBClient.DB(context.Background(), name)
+
+		if dbErr := db.Err(); dbErr != nil {
+			return nil, dbErr
+		}
+
+		return &CouchDBStore{db: db}, nil
+	}
+
+	err = p.couchDBClient.CreateDB(context.Background(), name)
+	if err != nil {
+		return nil, err
```

Comment:
There could be a race condition in which two clients are attempting to create the same DB. At the time you check if the DB exists it may not be there but then when you try to create it you'll get an error because it exists. You'll need to check again if the error means that the DB is there.

Reply:
The code has been changed so now there are separate create store and open store methods (per our discussion with @fqutishat on Friday), so the DB exists check is no longer there. Let me know if you still think there's a race condition.